### PR TITLE
REVM: Migration to 3.0

### PR DESCRIPTION
**Pull request by Evalir**
## Motivation

Advances the migration to REVM 3.0

Opening now to open a convo on the ongoing changes and make sure we're going in the right direction.

## Solution

Migrates types/imports to conform to 3.0. These changes are focused on _converting_ types—I'm deliberately choosing to keep the existing types that don't conflict there for now. The strategy is more or less as follows:

- All of the types that _need_ to use the revm types (e.g trait impls or structs/traits that interact very closely with revm) are switched.
- Everything else is just converted when needed (see `evm` crate `utils` for utility functions)

### CODE REVIEW

**Comment by Evalir** [comment]
i guess the main question i have here—do we wanna preserve some of the `ethers` types being used (`H256/U256` etc) or do we wanna fully switch to the revm primitives? happy to do either—just don't wanna potentially duplicate work and depending on the crate the revm types seem to be used vs the ethers ones.

i guess this depends if ethers will be eventually migrated to use ruint haha

**Comment by Evalir** [comment]
@mattsse we're down to around ~60-65 issues from ~450—feel free to start checking out the changes to know if you'd like to do them any other way (there's a lot of manual type conversion happening)

**Comment by Evalir** [comment]
aight @mattsse lil update—once we bump revm (i'm coordinating with dragan atm), this should compile. All issues have been solved—however of course we still need to run all tests and make sure types were converted properly.

Most, if not all type conversions should be abstracted into a few functions in the `evm/utils` crate (e.g `h256_to_b256`/`h160_to_b160`) and back. As you said, I just left the compiler guide me, so in places where `.into()` wasn't possible i used those (or just `.map()`ed it).

File: anvil/src/eth/backend/mem/mod.rs [hunk 1]
```diff
@@ -634,10 +648,28 @@ impl Backend {
...
+            }
+            ExecutionResult::Halt { reason, gas_used } => {
+                (halt_to_instruction_result(reason), gas_used, None, None)
+            }
+        };
+
         inspector.print_logs();
-        (exit_reason, out, gas_used, state, logs)
+
+        Ok((exit_reason, out, gas_used, state, logs.unwrap_or_default()))
```
Comment (Evalir) [inline anvil/src/eth/backend/mem/mod.rs:672]: highlighting this diff—one of the mayor logic changes with the new revm upgrade

File: anvil/src/eth/backend/executor.rs [hunk 2]
```diff
@@ -258,10 +267,30 @@ impl<'a, 'b, DB: Db + ?Sized, Validator: TransactionValidator> Iterator
...
+                (eval_to_instruction_result(reason), gas_used, Some(output), Some(logs))
+            }
+            ExecutionResult::Revert { gas_used, .. } => {
+                (InstructionResult::Revert, gas_used, None, None)
+            }
+            ExecutionResult::Halt { reason, gas_used } => {
+                (halt_to_instruction_result(reason), gas_used, None, None)
+            }
+        };
+
```
Comment (Evalir) [inline anvil/src/eth/backend/executor.rs:293]: ditto

File: anvil/src/eth/backend/mem/mod.rs [hunk 3]
```diff
@@ -932,16 +964,30 @@ impl Backend {
...
+            state.into_iter().map(|kv| (kv.0.into(), kv.1)).collect();
+        let (exit_reason, gas_used, out) = match result_and_state.result {
+            ExecutionResult::Success { reason, gas_used, output, .. } => {
+                (eval_to_instruction_result(reason), gas_used, Some(output))
+            }
+            ExecutionResult::Revert { gas_used, .. } => (InstructionResult::Revert, gas_used, None),
+            ExecutionResult::Halt { reason, gas_used } => {
+                (halt_to_instruction_result(reason), gas_used, None)
+            }
+        };
```
Comment (Evalir) [inline anvil/src/eth/backend/mem/mod.rs:990]: Ditto

File: anvil/src/eth/backend/mem/mod.rs [hunk 4]
```diff
@@ -959,8 +1005,22 @@ impl Backend {
...
+                ExecutionResult::Success { reason, gas_used, gas_refunded, logs, output } => {
+                    (eval_to_instruction_result(reason), gas_refunded, gas_used, Some(output), Some(logs))
+                },
+                ExecutionResult::Revert { gas_used, .. } => {
+                    (InstructionResult::Revert, 0 as u64, gas_used, None, None)
+                },
+                ExecutionResult::Halt { reason, gas_used } => {
+                    (halt_to_instruction_result(reason), 0 as u64, gas_used, None, None)
+                },
+            };
```
Comment (Evalir) [inline anvil/src/eth/backend/mem/mod.rs:1023]: Ditto

File: anvil/src/eth/backend/mem/mod.rs [hunk 5]
```diff
@@ -996,7 +1056,19 @@ impl Backend {
...
+        };
+        let (exit_reason, gas_used, out) = match result_and_state.result {
+            ExecutionResult::Success { reason, gas_used, output, .. } => {
+                (eval_to_instruction_result(reason), gas_used, Some(output))
+            }
+            ExecutionResult::Revert { gas_used, .. } => (InstructionResult::Revert, gas_used, None),
+            ExecutionResult::Halt { reason, gas_used } => {
+                (halt_to_instruction_result(reason), gas_used, None)
+            }
+        };
```
Comment (Evalir) [inline anvil/src/eth/backend/mem/mod.rs:1071]: Ditto

File: evm/src/executor/mod.rs [hunk 6]
```diff
@@ -753,15 +780,30 @@ fn calc_stipend(calldata: &[u8], spec: SpecId) -> u64 {
...
+        ExecutionResult::Success { reason, gas_used, gas_refunded, output, .. } => {
+            (eval_to_instruction_result(reason), gas_refunded, gas_used, Some(output))
+        }
+        ExecutionResult::Revert { gas_used, .. } => {
+            (InstructionResult::Revert, 0 as u64, gas_used, None)
+        }
+        ExecutionResult::Halt { reason, gas_used } => {
+            (halt_to_instruction_result(reason), 0 as u64, gas_used, None)
+        }
+    };
```
Comment (Evalir) [inline evm/src/executor/mod.rs:801]: Ditto

**Review by Evalir — COMMENTED** [review]
Highlighting some of the big differences in the migration here.

**Comment by mattsse** [comment]
sweet, sick work! 
could you please open this against the real foundry repo? then we can start the merge process!

**Comment by Evalir** [comment]
Awesome! Will do

**Pull request closed without merging**